### PR TITLE
Split source elements relative to their language.

### LINKED ISSUE #3068: [JS,TS] Range formatting ObjectExpression throws
Trying to format an ObjectExpression -- with at least 2 props -- throws. 

**Prettier master**

**Input:**
```sh
echo "const y = {a:1,b:2}" |  prettier --range-start 12 --range-end 14
```

**Output:**
```sh
[error] stdin: SyntaxError: Unexpected token, expected ; (1:7)
[error] > 1 | {a:1,b:2}
[error]     |       ^
```

**Expected behavior:**
Shouldn't throw and format the object as an object.
Note: if they were only 1 prop in the object it wouldn't throw but it would be considered as a `LabeledStatement`
```js
const y = {
     a: 1 ; // It would work with no-semi 
}
```
Other issue which seems highly related:
```sh
> echo "const y = [1]" |  prettier --range-start 11 --range-end 12
const y = [1;]
```

### PR DESCRIPTION
**Pull request by CiGit**
Fix #3068 : Colliding node types which are not source elements in every language.
Example: `ObjectExpression` in JSON / JS

Source Elements are now based on the parser.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -204,59 +204,73 @@ function isSourceElement(opts, node) {
...
+    "OperationTypeDefinition",
+    "InterfaceTypeDefinition",
+    "UnionTypeDefinition",
+    "ScalarTypeDefinition"
+  ];
+  switch (opts.parser) {
+    case "flow":
+    case "babylon":
+    case "typescript":
+      return jsSourceElements.includes(node.type);
```
Comment (CiGit) [inline index.js]: Let's end tests, but I'm not sure .includes is in every node ...

File: index.js [hunk 2]
```diff
@@ -204,59 +204,73 @@ function isSourceElement(opts, node) {
...
-    case "InputValueDefinition": // GraphQL
-    case "InputObjectTypeDefinition": // GraphQL
-    case "SchemaDefinition": // GraphQL
-    case "OperationTypeDefinition": // GraphQL
-    case "InterfaceTypeDefinition": // GraphQL
-    case "UnionTypeDefinition": // GraphQL
-    case "ScalarTypeDefinition": // GraphQL
-      return true;
+  // JS and JS like to avoid repetitions
+  const jsSourceElements = [
```
Comment (CiGit) [inline index.js]: What's the rule with global variables? those `const ...SourceElements` could be defined once.

File: tests/range/object-expression2.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+
+const y =       [
+    {
+                a: 1,
+    },<<<PRETTIER_RANGE_START>>>
+    {
+              a: 1,
```
Comment (azz) [inline tests/range/object-expression2.js]: Could you add an additional property to this object? Otherwise it could be treated as a block + label.

Comment (CiGit) [reply tests/range/object-expression2.js]: Right, it's certainly clearer than trying to find a semicolon `{a: 1;}`. 
Done.

File: index.js [hunk 4]
```diff
@@ -204,59 +204,73 @@ function isSourceElement(opts, node) {
...
-    case "EnumValueDefinition": // GraphQL
-    case "InputValueDefinition": // GraphQL
-    case "InputObjectTypeDefinition": // GraphQL
-    case "SchemaDefinition": // GraphQL
-    case "OperationTypeDefinition": // GraphQL
-    case "InterfaceTypeDefinition": // GraphQL
-    case "UnionTypeDefinition": // GraphQL
-    case "ScalarTypeDefinition": // GraphQL
-      return true;
+  // JS and JS like to avoid repetitions
```
Comment (josephfrazier) [inline index.js]: Is this supposed to be "JS and *TS* like to avoid repetitions"?

Comment (azz) [reply index.js]: I interpreted as `JS-like` = flow, TypeScript.

**Comment by josephfrazier** [comment]
Oh, that makes more sense. I had overlooked "like".

**Pull request merged**